src/main/java/edu/osu/waiting4ubackend/controller/UserController.java

```diff
@@ -76,5 +78,39 @@
         ObjectMapper objectMapper = new ObjectMapper();
         GetUserResponse getUserResponse = new GetUserResponse(user.getId(), user.getUserName(), user.getEmail(), user.getIntroduction(), user.getPreferences());
         return new ResponseEntity<>(objectMapper.writeValueAsString(getUserResponse), HttpStatus.OK);
-    }   
-}
\ No newline at end of file
+    }
+
+
+    @CrossOrigin
+    @PatchMapping(value = "/users/{id}", produces = "application/json")
+    public ResponseEntity<String> updateUser(@Valid @PathVariable long id, @RequestBody UserUpdateRequest request) throws IOException{
+        UserDBClient userDBClient = new UserDBClient();
+        User user = userDBClient.getUserById(id);
+
+        //check valid user id
+        if(user == null) {
+            return new ResponseEntity<>("{\"Error\":  \"The user doesn't exist\"}", HttpStatus.NOT_FOUND);
+        }
+
+        if(request.getUserName() != null){
+            //check duplicate userName
+           if(!request.getUserName().equals(user.getUserName()) & userDBClient.userNameExists(request.getUserName())) {
+                return new ResponseEntity<>("{\"Error\":  \"The name already exists, please use another one\"}", HttpStatus.FORBIDDEN);
+            }
+            user.setUserName(request.getUserName());
+        }
+
+        if(request.getPassword() != null){
+            user.setPassword(request.getPassword());
+        }
+
+        if(request.getIntroduction() != null){
+            user.setIntroduction(request.getIntroduction());
+        }
+
```

Comment:
How about the preference updating? I saw you implemented the updatePreference method.